kontrol/postgres.go

```diff
@@ -368,5 +376,94 @@ func insertQuery(kiteProt *protocol.Kite, url string) (string, []interface{}, er
 		"hostname",
 		"id",
 		"url",
+		"key_id",
 	).Values(values...).ToSql()
 }
+
+/*
+
+--- Key Pair -----------------
+
+*/
+
+func (p *Postgres) AddKey(keyPair *KeyPair) error {
+	if err := keyPair.Validate(); err != nil {
+		return err
+	}
+
+	psql := sq.StatementBuilder.PlaceholderFormat(sq.Dollar)
+	sqlQuery, args, err := psql.Insert("kite.key").Columns(
+		"id",
+		"public",
+		"private",
+	).Values(keyPair.ID, keyPair.Public, keyPair.Private).ToSql()
+	if err != nil {
+		return err
+	}
+
+	_, err = p.DB.Exec(sqlQuery, args...)
+	return err
+}
+
+func (p *Postgres) DeleteKey(keyPair *KeyPair) error {
+	res, err := p.DB.Exec(`UPDATE kite.key SET deleted_at = (now() at time zone 'utc')`)
+	if err != nil {
+		return err
+	}
+
+	n, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("n = %+v\n", n)
+	return nil
+}
+
+func (p *Postgres) GetKeyFromID(id string) (*KeyPair, error) {
+	psql := sq.StatementBuilder.PlaceholderFormat(sq.Dollar)
+	sqlQuery, args, err := psql.
+		Select("id", "public", "private").
+		From("kite.key").
+		Where(map[string]interface{}{"id": id}).
+		ToSql()
+	if err != nil {
+		return nil, err
+	}
+
+	keyPair := &KeyPair{}
+	err = p.DB.QueryRow(sqlQuery, args...).Scan(&keyPair.ID, &keyPair.Public, &keyPair.Private)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := keyPair.Validate(); err != nil {
+		return nil, err
+	}
+
+	return keyPair, nil
+}
+
+func (p *Postgres) GetKeyFromPublic(public string) (*KeyPair, error) {
+	psql := sq.StatementBuilder.PlaceholderFormat(sq.Dollar)
+	sqlQuery, args, err := psql.
+		Select("id", "public", "private").
+		From("kite.key").
+		Where(map[string]interface{}{"public": public}).
+		ToSql()
+	if err != nil {
+		return nil, err
+	}
+
+	keyPair := &KeyPair{}
+	err = p.DB.QueryRow(sqlQuery, args...).Scan(&keyPair.ID, &keyPair.Public, &keyPair.Private)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := keyPair.Validate(); err != nil {
+		return nil, err
+	}
+
+	return keyPair, nil
+}
```

Comment:
So you mean `GetKey*` functions should return an error when someone tries to retrieve an invalid key (keys that are not valid anymore, which in our Postgres implementation is marking it by changing the `deleted_at` row?